// react-lightning-design-system/lib/scripts/Table.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableRowColumnActions = exports.TableRowColumn = exports.TableHeaderColumn = exports.TableRow = exports.TableBody = exports.TableHeader = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _Icon = require('./Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TableHeader = exports.TableHeader = function (_Component) {
  (0, _inherits3.default)(TableHeader, _Component);

  function TableHeader() {
    (0, _classCallCheck3.default)(this, TableHeader);
    return (0, _possibleConstructorReturn3.default)(this, (TableHeader.__proto__ || (0, _getPrototypeOf2.default)(TableHeader)).apply(this, arguments));
  }

  (0, _createClass3.default)(TableHeader, [{
    key: 'renderBaseHeaderRow',
    value: function renderBaseHeaderRow() {
      var _props = this.props,
          children = _props.children,
          sortable = _props.sortable,
          hasActions = _props.hasActions,
          actionsPosition = _props.actionsPosition;

      var nextChildren = [];

      var props = {
        className: 'slds-text-title--caps'
      };

      _react2.default.Children.forEach(children.props.children, function (child, index) {
        var childSortable = child.props.sortable;
        nextChildren.push(_react2.default.cloneElement(child, {
          key: index,
          sortable: typeof childSortable === 'undefined' ? sortable : childSortable
        }));
      });

      if (hasActions) {
        nextChildren = [].concat((0, _toConsumableArray3.default)(nextChildren.slice(0, actionsPosition)), [_react2.default.createElement(TableHeaderColumn, {
          sortable: false,
          width: 50,
          key: -1,
          className: 'slds-cell-shrink'
        })], (0, _toConsumableArray3.default)(nextChildren.slice(actionsPosition)));
      }

      return _react2.default.cloneElement(children, props, nextChildren);
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'thead',
        null,
        this.renderBaseHeaderRow()
      );
    }
  }]);
  return TableHeader;
}(_react.Component);

TableHeader.propTypes = {
  hasActions: _react.PropTypes.bool,
  actionsPosition: _react.PropTypes.number,
  children: _react.PropTypes.node,
  sortable: _react.PropTypes.bool
};

TableHeader.defaultProps = {
  actionsPosition: 0
};

var TableBody = exports.TableBody = function (_Component2) {
  (0, _inherits3.default)(TableBody, _Component2);

  function TableBody() {
    (0, _classCallCheck3.default)(this, TableBody);
    return (0, _possibleConstructorReturn3.default)(this, (TableBody.__proto__ || (0, _getPrototypeOf2.default)(TableBody)).apply(this, arguments));
  }

  (0, _createClass3.default)(TableBody, [{
    key: 'renderRows',
    value: function renderRows() {
      return _react2.default.Children.map(this.props.children, function (child) {
        var children = [];

        _react2.default.Children.forEach(child.props.children, function (innerChild, index) {
          if (!_react2.default.isValidElement(innerChild)) return;
          var truncate = innerChild.props.truncate;

          var props = {
            key: index
          };
          if (typeof truncate !== 'undefined') props.truncate = truncate;
          children.push(_react2.default.cloneElement(innerChild, props));
        });

        return _react2.default.cloneElement(child, { className: 'slds-hint-parent' }, children);
      });
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'tbody',
        null,
        this.renderRows()
      );
    }
  }]);
  return TableBody;
}(_react.Component);

TableBody.propTypes = {
  children: _react.PropTypes.node
};

var TableRow = function TableRow(_ref) {
  var selected = _ref.selected,
      props = (0, _objectWithoutProperties3.default)(_ref, ['selected']);
  var style = props.style;


  if (selected) {
    style = (0, _assign2.default)({}, style, {
      backgroundColor: '#F8FCF5',
      borderLeft: '2px solid #7db450'
    });
  }

  return _react2.default.createElement(
    'tr',
    (0, _extends3.default)({}, props, { style: style }),
    props.children
  );
};

exports.TableRow = TableRow;
TableRow.propTypes = {
  selected: _react.PropTypes.bool,
  children: _react.PropTypes.node,
  /* eslint-disable react/forbid-prop-types */
  style: _react.PropTypes.object
};

var TableHeaderColumn = function TableHeaderColumn(props) {
  var sortable = props.sortable,
      resizable = props.resizable,
      children = props.children,
      className = props.className,
      width = props.width,
      sortDir = props.sortDir,
      onSort = props.onSort,
      sorted = props.sorted,
      align = props.align,
      pprops = (0, _objectWithoutProperties3.default)(props, ['sortable', 'resizable', 'children', 'className', 'width', 'sortDir', 'onSort', 'sorted', 'align']);

  var oClassNames = (0, _classnames3.default)(className, 'slds-text-title--caps slds-truncate', (0, _defineProperty3.default)({
    'slds-is-sortable': sortable,
    'slds-is-resizable': resizable
  }, 'slds-text-align--' + align, align));

  var style = { minWidth: width || 'auto' };

  var buttonStyle = {};

  // TODO: SLDS doesn't have class for selected so will wait
  if (sorted) {
    style.backgroundColor = '#F4F6F9';
    style.color = '#0070d2';
    buttonStyle.visibility = 'visible';
  }

  var icon = sortDir === 'DESC' ? 'arrowdown' : 'arrowup';

  return _react2.default.createElement(
    'th',
    (0, _extends3.default)({}, pprops, {
      className: oClassNames,
      scope: 'col',
      style: style
    }),
    sortable ? _react2.default.createElement(
      'a',
      {
        onClick: function onClick(e) {
          e.preventDefault();onSort();
        },
        className: 'slds-th__action slds-text-link--reset'
      },
      _react2.default.createElement(
        'span',
        { className: 'slds-assistive-text' },
        'Sort '
      ),
      _react2.default.createElement(
        'span',
        { className: 'slds-truncate' },
        children
      ),
      _react2.default.createElement(_Icon2.default, {
        className: 'slds-is-sortable__icon',
        textColor: 'default',
        container: true,
        size: 'x-small',
        category: 'utility',
        icon: icon,
        style: { position: 'absolute' }
      }),
      _react2.default.createElement('span', { className: 'slds-assistive-text', 'aria-live': 'assertive', 'aria-atomic': 'true' })
    ) : children
  );
};

exports.TableHeaderColumn = TableHeaderColumn;
TableHeaderColumn.propTypes = {
  className: _react.PropTypes.string,
  onSort: _react.PropTypes.func,
  width: _react.PropTypes.oneOfType([_react2.default.PropTypes.string, _react2.default.PropTypes.number]),
  sortable: _react.PropTypes.bool,
  resizable: _react.PropTypes.bool,
  sortDir: _react.PropTypes.string,
  sorted: _react.PropTypes.bool,
  align: _react.PropTypes.oneOf(['left', 'center', 'right']),
  children: _react.PropTypes.node
};

var TableRowColumn = function TableRowColumn(props) {
  var truncate = props.truncate,
      className = props.className,
      children = props.children,
      pprops = (0, _objectWithoutProperties3.default)(props, ['truncate', 'className', 'children']);

  var oClassNames = (0, _classnames3.default)(className, {
    'slds-truncate': truncate
  });
  var style = {};
  if (!truncate) style.position = 'static';

  return _react2.default.createElement(
    'td',
    (0, _extends3.default)({
      role: 'gridcell',
      style: style,
      className: oClassNames
    }, pprops),
    children
  );
};

exports.TableRowColumn = TableRowColumn;
TableRowColumn.propTypes = {
  className: _react.PropTypes.string,
  truncate: _react.PropTypes.bool,
  children: _react.PropTypes.node
};

TableRowColumn.defaultProps = {
  truncate: true
};

var TableRowColumnActions = exports.TableRowColumnActions = function TableRowColumnActions(props) {
  return _react2.default.createElement(
    TableRowColumn,
    {
      className: 'slds-cell-shrink',
      'data-label': 'Actions',
      truncate: false,
      width: 50,
      style: { position: 'static' }
    },
    props.children
  );
};

TableRowColumnActions.propTypes = {
  children: _react.PropTypes.node
};

var Table = function (_Component3) {
  (0, _inherits3.default)(Table, _Component3);

  function Table() {
    (0, _classCallCheck3.default)(this, Table);
    return (0, _possibleConstructorReturn3.default)(this, (Table.__proto__ || (0, _getPrototypeOf2.default)(Table)).apply(this, arguments));
  }

  (0, _createClass3.default)(Table, [{
    key: 'onScroll',
    value: function onScroll() {
      var elements = document.getElementsByClassName('react-slds-dropdown-opened');
      if (elements.length) elements[0].childNodes[0].blur();
    }
  }, {
    key: 'renderTableHeader',
    value: function renderTableHeader(base) {
      var sortable = this.props.sortable;

      return _react2.default.cloneElement(base, { sortable: sortable });
    }
  }, {
    key: 'renderTableBody',
    value: function renderTableBody(base) {
      return base;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this4 = this;

      var _props2 = this.props,
          className = _props2.className,
          bordered = _props2.bordered,
          verticalBorders = _props2.verticalBorders,
          noRowHover = _props2.noRowHover,
          striped = _props2.striped,
          fixedLayout = _props2.fixedLayout,
          children = _props2.children,
          autoWidth = _props2.autoWidth,
          wrapperStyle = _props2.wrapperStyle,
          pprops = (0, _objectWithoutProperties3.default)(_props2, ['className', 'bordered', 'verticalBorders', 'noRowHover', 'striped', 'fixedLayout', 'children', 'autoWidth', 'wrapperStyle']);

      delete pprops.sortable;

      var tableClassNames = (0, _classnames3.default)(className, 'slds-table slds-table--cell-buffer', {
        'slds-table--bordered': bordered,
        'slds-no-row-hover': noRowHover,
        'slds-table--striped': striped,
        'slds-table--fixed-layout': fixedLayout,
        'slds-table--col-bordered': verticalBorders
      });

      var wrapStyle = (0, _assign2.default)({
        overflowY: 'hidden',
        overflowX: 'auto'
      }, wrapperStyle);

      var style = {};
      if (autoWidth) style.width = 'auto';

      var tBody = void 0;
      var tHead = void 0;

      _react2.default.Children.forEach(children, function (child) {
        if (!_react2.default.isValidElement(child)) return;
        if (child.type === TableHeader) {
          tHead = _this4.renderTableHeader(child);
        } else if (child.type === TableBody) {
          tBody = _this4.renderTableBody(child);
        }
      });

      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          'div',
          { style: wrapStyle, onScroll: this.onScroll.bind(this) },
          _react2.default.createElement(
            'table',
            (0, _extends3.default)({ className: tableClassNames, style: style }, pprops),
            tHead,
            tBody
          )
        )
      );
    }
  }]);
  return Table;
}(_react.Component);

Table.propTypes = {
  wrapperStyle: _react.PropTypes.object,
  className: _react.PropTypes.string,
  bordered: _react.PropTypes.bool,
  verticalBorders: _react.PropTypes.bool,
  noRowHover: _react.PropTypes.bool,
  striped: _react.PropTypes.bool,
  fixedLayout: _react.PropTypes.bool,
  sortable: _react.PropTypes.bool,
  children: _react.PropTypes.node,
  autoWidth: _react.PropTypes.bool
};

exports.default = Table;
